### PR TITLE
Implement BIP-175 address

### PR DESCRIPTION
**Pull request by Yamaguchi**
This PR implements "Pay to Contract address" defined in  BIP-175

see https://github.com/bitcoin/bips/blob/master/bip-0175.mediawiki

### CODE REVIEW

File: lib/tapyrus/bip175.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    PURPOSE_TYPE = 175
+
+    attr_accessor :payment_base
+
+    def initialize
+      @contracts = []
+    end
+
+    # @param key [Tapyrus::ExtKey] master private extended key
+    def self.from_private_key(key)
```
Comment (azuchi) [inline lib/tapyrus/bip175.rb:24]: `from_ext_key` is better because it clarify `key` is Extkey. And please add validation whether key is_a?(ExtKey).

File: lib/tapyrus/bip175.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      new.tap do |bip175|
+        bip175.payment_base =
+          key.derive(PURPOSE_TYPE, true).derive(Tapyrus.chain_params.bip44_coin_type, true).ext_pubkey
+      end
+    end
+
+    # @param key [Tapyrus::ExtPubkey] contract base public key
+    def self.from_public_key(key)
```
Comment (azuchi) [inline lib/tapyrus/bip175.rb:32]: `from_ext_pubkey` is better because it clarify `key` is ExtPubkey. And please add validation whether key is_a?(ExtPubkey)

File: spec/tapyrus/bip175_spec.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    subject { key.addr }
+
+    before do
+      key << document1
+      key << document2
+    end
+    let(:key) { Tapyrus::BIP175.from_private_key(master) }
+
+    it { is_expected.to eq '1C7f322izqMqLzZzfzkPAjxBzprxDi47Yf' }
+  end
```
Comment (azuchi) [inline spec/tapyrus/bip175_spec.rb:63]: Please add test defined by BIP175(in `Verification example (negative test)` section).

**Review by azuchi — CHANGES_REQUESTED** [review]

File: lib/tapyrus/bip175.rb (hunk 1 above)
Comment (Yamaguchi) [reply lib/tapyrus/bip175.rb]: Rename method name (fb72383)

File: lib/tapyrus/bip175.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    PURPOSE_TYPE = 175
+
+    attr_accessor :payment_base
+
+    def initialize
+      @contracts = []
+    end
+
+    # @param key [Tapyrus::ExtKey] master private extended key
+    def self.from_ext_key(key)
```
Comment (azuchi) [inline lib/tapyrus/bip175.rb:24]: If you provide this method, you need to verify that the argument key is the master private extended key.
If it is not the master private extended key, you will get the wrong key derivation.

**Review by azuchi — CHANGES_REQUESTED** [review]

File: lib/tapyrus/bip175.rb (hunk 4 above)
Comment (azuchi) [reply lib/tapyrus/bip175.rb:24]: It is good that ExtKey itself has the judgment of whether it is a master or not.

File: lib/tapyrus/ext_key.rb [hunk 5]
```diff
@@ -356,5 +360,9 @@ def self.support_version?(version)
       p = Tapyrus.chain_params
       [p.bip49_pubkey_p2wpkh_p2sh_version, p.bip84_pubkey_p2wpkh_version, p.extended_pubkey_version].include?(version)
     end
+
+    def master?
+      depth == 0
```
Comment (azuchi) [inline lib/tapyrus/ext_key.rb:365]: In addition to `depth`, you also need to check number and fingerprint.

**Review by azuchi — CHANGES_REQUESTED** [review]

File: lib/tapyrus/ext_key.rb (hunk 5 above)
Comment (Yamaguchi) [reply lib/tapyrus/ext_key.rb]: Add check against number and fingerprint(97396b6)

**Review by azuchi — APPROVED** [review]
